Objective-C/CBLDatabase.mm

```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
         
-        CBLDocument* resolvedDoc = [self resolveConflictBetweenLocalDoc: localDoc
-                                                           andRemoteDoc: remoteDoc];
+        NSError* err;
+        BOOL success = [self saveResolvedDocument: resolvedDoc withLocalDoc: localDoc
+                                        remoteDoc: remoteDoc error: &err];
         
-        // Save resolved document:
-        if (![self saveResolvedDocument: resolvedDoc
-                           withLocalDoc: localDoc
-                              remoteDoc: remoteDoc
-                                  error: outError])
-            return false;
+        if ($equal(err.domain, CBLErrorDomain) && err.code == CBLErrorConflict)
+            continue;
         
-        return t.commit() || convertError(t.error(), outError);
+        if (outError)
+            *outError = err;
+        
+        return success;
     }
 }
 
-
-- (CBLDocument*) resolveConflictBetweenLocalDoc: (CBLDocument*)localDoc
-                                   andRemoteDoc: (CBLDocument*)remoteDoc {
-    if (remoteDoc.isDeleted)
-        return remoteDoc;
-    else if (localDoc.isDeleted)
-        return localDoc;
-    else if (localDoc.generation > remoteDoc.generation)
-        return localDoc;
-    else if (localDoc.generation < remoteDoc.generation)
-        return remoteDoc;
-    else if ([localDoc.revID compare: remoteDoc.revID] > 0)
-        return localDoc;
-    else
-        return remoteDoc;
-}
-
-
 - (BOOL) saveResolvedDocument: (CBLDocument*)resolvedDoc
                  withLocalDoc: (CBLDocument*)localDoc
                     remoteDoc: (CBLDocument*)remoteDoc
                         error: (NSError**)outError
 {
-    if (resolvedDoc != localDoc)
-        resolvedDoc.database = self;
-    
-    // The remote branch has to win, so that the doc revision history matches the server's.
-    CBLStringBytes winningRevID = remoteDoc.revID;
-    CBLStringBytes losingRevID = localDoc.revID;
-    
-    alloc_slice mergedBody;
-    C4RevisionFlags mergedFlags = 0;
-    if (resolvedDoc != remoteDoc) {
-        // Unless the remote revision is being used as-is, we need a new revision:
-        mergedBody = [resolvedDoc encode: outError];
-        if (!mergedBody)
-            return false;
-        if (resolvedDoc.isDeleted)
-            mergedFlags |= kRevDeleted;
-    }
-    
-    // Tell LiteCore to do the resolution:
-    C4Document *rawDoc = localDoc.c4Doc.rawDoc;
-    C4Error c4err;
-    if (!c4doc_resolveConflict(rawDoc,
-                               winningRevID,
-                               losingRevID,
-                               mergedBody,
-                               mergedFlags,
-                               &c4err)
-        || !c4doc_save(rawDoc, 0, &c4err)) {
-        return convertError(c4err, outError);
+    CBL_LOCK(self) {
+        C4Transaction t(_c4db);
+        if (!t.begin())
+            return convertError(t.error(), outError);
+        
```

Comment:
If resolvedDoc is nil, we could try to match it to localDoc or remoteDoc only if one of them is a deletedDoc here.